internal/jobworker/grpc/server.go

```diff
@@ -0,0 +1,230 @@
+	userSvc IUserService
+}
+
+func (jw JobWorker) Start(ctx context.Context, req *pb.StartRequest) (*pb.StartResponse, error) {
+	user, ok := jw.userSvc.User(ctx)
+	if !ok {
+		return nil, status.Error(codes.Unauthenticated, "unauthenticated")
+	}
+
+	valid := validator.New()
+	valid.AssertFunc(func() bool { return req.Command != nil }, "command empty")
+	valid.AssertFunc(func() bool { return req.Command.Name != "" }, "command name empty")
+	valid.AssertFunc(func() bool { return req.Limits != nil }, "limits empty")
+	if err := valid.Err(); err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	logger.Infof("processing StartRequest; Command: %v", req.Command)
+
+	j, err := job.New(
+		user,
+		reexec.Command{
+			Name: req.Command.Name,
+			Args: req.Command.Args,
+		},
+	)
+	if err != nil {
+		logger.Errorf("building Job; error: %v", err)
+		return nil, status.Error(codes.Internal, "error building job")
+	}
+
+	if err := jw.jobSvc.StartJob(
+		ctx,
+		*j,
+		cgroupOptions(req.Limits)...,
+	); err != nil {
+		logger.Errorf("starting Job; error: %v", err)
+		return nil, status.Error(codes.Internal, "error starting job")
+	}
+
+	logger.Infof("Job started; ID: %v", j.ID)
+	return &pb.StartResponse{
+		JobId:   j.ID.String(),
+		Command: req.Command,
+		Status: &pb.StatusDetail{
+			Status:   toStatus(j.Status()),
+			ExitCode: int32(j.ExitCode()),
+		},
+		Limits: req.Limits,
+	}, nil
+}
+
+func (jw JobWorker) Stop(ctx context.Context, req *pb.StopRequest) (*pb.StopResponse, error) {
+	user, ok := jw.userSvc.User(ctx)
+	if !ok {
+		return nil, status.Error(codes.Unauthenticated, "unauthenticated")
+	}
+
+	if req.JobId == "" {
+		return nil, status.Error(codes.InvalidArgument, validator.Format("empty job ID"))
+	}
+
+	j, err := jw.fetchJob(ctx, user, req.JobId)
+	if err != nil {
+		return nil, err
+	}
+
+	if j.Status() != job.Running {
+		return nil, status.Error(codes.FailedPrecondition, "job is not running")
+	}
+
+	if err := jw.jobSvc.StopJob(ctx, j.ID); err != nil {
+		logger.Errorf("stop job; job: %s, error: %v", j.ID, err)
+		return nil, status.Error(codes.Internal, "error stopping job")
+	}
+
+	return &pb.StopResponse{}, nil
+}
+
+func (jw JobWorker) Status(ctx context.Context, req *pb.StatusRequest) (*pb.StatusResponse, error) {
+	user, ok := jw.userSvc.User(ctx)
+	if !ok {
+		return nil, status.Error(codes.Unauthenticated, "unauthenticated")
+	}
+
+	if req.JobId == "" {
+		return nil, status.Error(codes.InvalidArgument, validator.Format("empty job ID"))
+	}
+
+	j, err := jw.fetchJob(ctx, user, req.JobId)
+	if err != nil {
+		return nil, err
+	}
+
+	return &pb.StatusResponse{
+		Status: &pb.StatusDetail{
+			Status:   toStatus(j.Status()),
+			ExitCode: int32(j.ExitCode()),
+		},
+	}, nil
+}
+
+func (jw JobWorker) Output(req *pb.OutputRequest, stream pb.JobWorkerService_OutputServer) error {
+	user, ok := jw.userSvc.User(stream.Context())
+	if !ok {
+		return status.Error(codes.Unauthenticated, "unauthenticated")
```

Comment:
A gRPC filter could solve some of this repetition, plus it centers authz in a single place and applies to (potentially) all RPCs.